### PR TITLE
axp - Chapter 9 - The Composite Pattern

### PR DESCRIPTION
**Pull request by a-panayotopoulos**
This chapter covers the Iterator and Composite patterns. To demonstrate the Composite pattern I've created a tree of organisms, with methods to add, remove and query them.

I've also implemented methods to iterate through the tree. Rather than implementing any Java `Iterator` class, though, I've gone with the Java 8 concept of a Stream. I'd argue this is essentially still the Iterator pattern, albeit with what the following article calls a "passive iterator":
http://www.javaworld.com/article/2461744/java-language/java-language-iterating-over-collections-in-java-8.html

### CODE REVIEW

File: axp-test/org/axp/composite/CompositeTest.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+		assertTrue( "Did not find all organisms", expected.isEmpty() );
+	}
+	
+	@Test
+	public void testHeavier() {
+		Queue<String> expected = queueOf( "cat", "lion" );
+		OrganismTree duck = organisms.find( "duck" );
+		
+		organisms.stream()
+			.filter( critter -> critter.heavierThan( duck.getAvgWeight() ) )
```
Comment (txr12) [inline axp-test/org/axp/composite/CompositeTest.java]: criter ? criteria? criterion  </pedantic>

File: axp-test/org/axp/composite/CompositeTest.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	@Test public void testStream() {
+		// Show first item is the root of the tree
+		assertEquals( "Organism", organisms.stream().findFirst().get().getName() );
+		
+		// Next show a full list of items, sorted by name
+		Queue<String> expected = queueOf(
+				"Animal", "Arthropod", "Bird", "Carp", "Feline", "Fish", "Mammal", "Organism", "Rodent", "Spider", "Waterfowl",
+				"cat", "cellar spider", "duck", "goldfish", "lion", "mouse" );
+		
+		organisms.stream()
```
Comment (txr12) [inline axp-test/org/axp/composite/CompositeTest.java]: need to look at streams (and java 8 altogether i guess)

File: axp/org/axp/composite/OrganismTree.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	
+	/**
+	 * Is this organism lighter than a particular weight, on average?
+	 * @param weight a weight in kilos
+	 * @return true iff this is an organism with a defined average weight which is lighter than the figure given
+	 */
+	public boolean lighterThan( float weight ) {
+		return ( ( getAvgWeight() != null ) && ( getAvgWeight() < weight ) );
+	}
+}
```
Comment (txr12) [inline axp/org/axp/composite/OrganismTree.java]: great example. really a nice idea for composite

File: axp-test/org/axp/composite/CompositeTest.java (hunk 1 above)
Comment (a-panayotopoulos) [inline axp-test/org/axp/composite/CompositeTest.java]: [**critter** (usually affectionate) A creature, an animal.](https://en.wiktionary.org/wiki/critter) :dog:

File: axp-test/org/axp/composite/CompositeTest.java (hunk 2 above)
Comment (a-panayotopoulos) [inline axp-test/org/axp/composite/CompositeTest.java]: I recommend giving it an hour or so to read up and then have a quick play. They're pretty powerful.

File: axp/org/axp/composite/OrganismTree.java (hunk 3 above)
Comment (a-panayotopoulos) [inline axp/org/axp/composite/OrganismTree.java]: Thanks!

File: axp-test/org/axp/composite/CompositeTest.java (hunk 1 above)
Comment (txr12) [inline axp-test/org/axp/composite/CompositeTest.java]: yup, it makes sense. My english left me when i wrote that comment.

**Pull request merged**